diff --git a/magic/Magdir/msdos b/magic/Magdir/msdos
index 43c1d97..f0578f4 100644
--- a/magic/Magdir/msdos
+++ b/magic/Magdir/msdos
@@ -682,6 +682,16 @@
 
 0	string	QEVM		QEMU's suspend to disk image
 
+# QEMU QED Image
+# http://wiki.qemu.org/Features/QED/Specification
+0	string/b	QED\0		QEMU QED Image
+
+# VDI Image
+64	string/b	\x7f\x10\xda\xbe	VDI Image
+>68	string/b	\x01\x00\x01\x00	version 1.1
+>0	string		>\0			(%s)
+>368	lequad		x			 \b, %lld bytes
+
 0	string	Bochs\ Virtual\ HD\ Image	Bochs disk image,
 >32	string	x				type %s,
 >48	string	x				subtype %s
