diff --git a/doc/file.man b/doc/file.man
index 444842a..06dc4c0 100644
--- a/doc/file.man
+++ b/doc/file.man
@@ -545,6 +545,11 @@ This manual page, and particularly this section, is too long.
 .Sh RETURN CODE
 .Nm
 returns 0 on success, and non-zero on error.
+.Pp
+If the file named by the file
+operand does not exist, cannot be read, or the type of the file named by
+the file operand cannot be determined, this is not be considered an error
+that affects the exit status.
 .Sh AVAILABILITY
 You can obtain the original author's latest version by anonymous FTP
 on
