diff --git a/doc/magic.man b/doc/magic.man
index f95280c..0a7e851 100644
--- a/doc/magic.man
+++ b/doc/magic.man
@@ -71,7 +71,7 @@ characters in the magic match both lower and upper case characters in the
 target, whereas upper case characters in the magic only match uppercase
 characters in the target.
 .It Dv pstring
-A Pascal-style string where the first byte is interpreted as the an
+A Pascal-style string where the first byte is interpreted as the
 unsigned length.
 The string is not NUL terminated.
 .It Dv date
@@ -348,7 +348,7 @@ If a the test on a line at level
 .Em n
 succeeds, all following tests at level
 .Em n+1
-are performed, and the messages printed if the tests succeed, untile a line
+are performed, and the messages printed if the tests succeed, until a line
 with level
 .Em n
 (or less) appears.
