diff --git a/src/magic.c b/src/magic.c
index d49c29a..c2a95ae 100644
--- a/src/magic.c
+++ b/src/magic.c
@@ -85,6 +85,7 @@ private const char *
 get_default_magic(void)
 {
 	static const char hmagic[] = "/.magic";
+	static const char hmagic_mgc[] = "/.magic.mgc";
 	static char default_magic[2 * MAXPATHLEN + 2];
 	char *home;
 	char hmagicpath[MAXPATHLEN + 1];
@@ -92,10 +93,14 @@ get_default_magic(void)
 	if ((home = getenv("HOME")) == NULL)
 		return MAGIC;
 
-	(void)snprintf(hmagicpath, sizeof(hmagicpath), "%s%s", home, hmagic);
+	(void)snprintf(hmagicpath, sizeof(hmagicpath), "%s%s", home, hmagic_mgc);
 
-	if (access(hmagicpath, R_OK) == -1)
-		return MAGIC;
+	if (access(hmagicpath, R_OK) == -1) {
+		(void)snprintf(hmagicpath, sizeof(hmagicpath), "%s%s", home, hmagic);
+		if (access(hmagicpath, R_OK) == -1) {
+			return MAGIC;
+		}
+	}
 
 	(void)snprintf(default_magic, sizeof(default_magic), "%s:%s",
 	    hmagicpath, MAGIC);
