diff --git a/magic/Magdir/sgml b/magic/Magdir/sgml
index 8d97c1e..63248a0 100644
--- a/magic/Magdir/sgml
+++ b/magic/Magdir/sgml
@@ -16,17 +16,17 @@
 !:mime	application/xml-sitemap
 
 # xhtml
-0	string		\<?xml\ version="
->15	string		>\0
->>19	search/4096/cWbt	\<!doctype\ html	xHTML document text
+0	string/t		\<?xml\ version="
+>19	search/4096/cWbt	\<!doctype\ html	XHTML document text
+>>15	string		>\0	(version %.3s)
 !:mime	text/html
-0	string		\<?xml\ version='
->15	string		>\0
->>19	search/4096/cWbt	\<!doctype\ html	xHTML document text
+0	string/t		\<?xml\ version='
+>19	search/4096/cWbt	\<!doctype\ html	XHTML document text
+>>15	string		>\0	(version %.3s)
 !:mime	text/html
-0	string		\<?xml\ version="
->15	string		>\0
->>19	search/4096/cWbt	\<html	broken xHTML document text
+0	string/t		\<?xml\ version="
+>19	search/4096/cWbt	\<html	broken XHTML document text
+>>15	string		>\0	(version %.3s)
 !:mime	text/html
 
 #------------------------------------------------------------------------------
@@ -45,24 +45,21 @@
 
 # Extensible markup language (XML), a subset of SGML
 # from Marc Prud'hommeaux (marc@apocalypse.org)
-0	search/1/cwbt	\<?xml			XML document text
+0	search/1/cwt	\<?xml			XML document text
 !:mime	application/xml
-0	string		\<?xml\ version\ "	XML
+0	string/t		\<?xml\ version\ "	XML
 !:mime	application/xml
-0	string		\<?xml\ version="	XML
+0	string/t		\<?xml\ version="	XML
 !:mime	application/xml
->15	search/1	>\0			%.3s document text
+>15	string/t	>\0			%.3s document text
 >>23	search/1	\<xsl:stylesheet	(XSL stylesheet)
 >>24	search/1	\<xsl:stylesheet	(XSL stylesheet)
 0	string		\<?xml\ version='	XML
 !:mime	application/xml
->15	search/1	>\0			%.3s document text
+>15	string/t	>\0			%.3s document text
 >>23	search/1	\<xsl:stylesheet	(XSL stylesheet)
 >>24	search/1	\<xsl:stylesheet	(XSL stylesheet)
-0	search/1/wbt	\<?xml			XML document text
-!:mime	application/xml
-!:strength - 10
-0	search/1/wbt	\<?XML			broken XML document text
+0	search/1/wt	\<?XML			broken XML document text
 !:mime	application/xml
 !:strength - 10
 
