diff --git a/magic/Magdir/compress b/magic/Magdir/compress
index d529486..7158f77 100644
--- a/magic/Magdir/compress
+++ b/magic/Magdir/compress
@@ -196,6 +196,13 @@
 # bug #364260)
 #0	string		]\000\000\200\000	LZMA compressed data
 
+# Type: LZMA
+0	lelong&0xffffff	=0x5d
+>12	leshort		=0xff			LZMA compressed data,
+>>5	lequad		=0xffffffffffffffff	streamed
+>>5	lequad		!0xffffffffffffffff	non-streamed, size %lld
+!:mime	application/x-lzma
+
 # http://tukaani.org/xz/xz-file-format.txt
 0	ustring		\xFD7zXZ\x00		xz compressed data
 !:mime	application/x-xz
