diff --git a/magic/Magdir/linux b/magic/Magdir/linux
index 4cbf0fb..a95fa9f 100644
--- a/magic/Magdir/linux
+++ b/magic/Magdir/linux
@@ -66,6 +66,7 @@
 # From Daniel Novotny <dnovotny@redhat.com>
 # swap file for PowerPC
 65526	string		SWAPSPACE2	Linux/ppc swap file
+16374   string          SWAPSPACE2      Linux/ia64 swap file
 # ECOFF magic for OSF/1 and Linux (only tested under Linux though)
 #
 #	from Erik Troan (ewt@redhat.com) examining od dumps, so this
